File: app/src/main/java/fr/nihilus/mymusic/service/MediaNotificationManager.java
package fr.nihilus.mymusic.service;

import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Build;
import android.os.RemoteException;
import android.support.annotation.DrawableRes;
import android.support.v4.media.MediaDescriptionCompat;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaControllerCompat;
import android.support.v4.media.session.MediaSessionCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.support.v7.app.NotificationCompat;
import android.support.v7.content.res.AppCompatResources;
import android.util.Log;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.request.animation.GlideAnimation;
import com.bumptech.glide.request.target.SimpleTarget;

import fr.nihilus.mymusic.HomeActivity;
import fr.nihilus.mymusic.R;
import fr.nihilus.mymusic.utils.ResourceHelper;

/**
 * Une classe associée à {@link MusicService} permettant l'affichage d'une notification
 * pour contrôler le Playback.
 */
public class MediaNotificationManager extends BroadcastReceiver {

    private static final int NOTIFICATION_ID = 16;
    private static final int REQUEST_CODE = 100;
    private static final String ACTION_PAUSE = "fr.nihilus.nihilusmusic.PAUSE";
    private static final String ACTION_PLAY = "fr.nihilus.nihilusmusic.PLAY";
    private static final String ACTION_PREV = "fr.nihilus.nihilusmusic.PREV";
    private static final String ACTION_NEXT = "fr.nihilus.nihilusmusic.NEXT";
    private static final String TAG = "NotificationManager";
    private final MusicService mService;
    private final NotificationManager mNotificationManager;
    private final PendingIntent mPauseIntent;
    private final PendingIntent mPlayIntent;
    private final PendingIntent mPreviousIntent;
    private final PendingIntent mNextIntent;
    private MediaSessionCompat.Token mSessionToken;
    private MediaControllerCompat mController;
    private MediaControllerCompat.TransportControls mTransportControls;
    private PlaybackStateCompat mPlaybackState;
    private MediaMetadataCompat mMetadata;
    private final Bitmap mDummyAlbumArt;
    private final int mAccentColor;
    private boolean mStarted = false;

    private final MediaControllerCompat.Callback mCallback = new MediaControllerCompat.Callback() {
        @Override
        public void onPlaybackStateChanged(PlaybackStateCompat state) {
            mPlaybackState = state;
            if (state != null && (state.getState() == PlaybackStateCompat.STATE_STOPPED
                    || state.getState() == PlaybackStateCompat.STATE_NONE)) {
                stopNotification();
            } else {
                Notification notification = createNotification();
                if (notification != null) {
                    mNotificationManager.notify(NOTIFICATION_ID, notification);
                }
            }
        }

        @Override
        public void onMetadataChanged(MediaMetadataCompat metadata) {
            mMetadata = metadata;
            Log.d(TAG, "Received new metadata " + metadata);
            Notification notification = createNotification();
            if (notification != null) {
                mNotificationManager.notify(NOTIFICATION_ID, notification);
            }
        }

        @Override
        public void onSessionDestroyed() {
            updateSessionToken();
        }
    };

    public MediaNotificationManager(MusicService service) {
        mService = service;
        updateSessionToken();

        mNotificationManager = (NotificationManager)
                mService.getSystemService(Context.NOTIFICATION_SERVICE);
        String pkg = mService.getPackageName();
        mPauseIntent = PendingIntent.getBroadcast(mService, REQUEST_CODE,
                new Intent(ACTION_PAUSE).setPackage(pkg),
                PendingIntent.FLAG_CANCEL_CURRENT);
        mPlayIntent = PendingIntent.getBroadcast(mService, REQUEST_CODE,
                new Intent(ACTION_PLAY).setPackage(pkg),
                PendingIntent.FLAG_CANCEL_CURRENT);
        mPreviousIntent = PendingIntent.getBroadcast(mService, REQUEST_CODE,
                new Intent(ACTION_PREV).setPackage(pkg),
                PendingIntent.FLAG_CANCEL_CURRENT);
        mNextIntent = PendingIntent.getBroadcast(mService, REQUEST_CODE,
                new Intent(ACTION_NEXT).setPackage(pkg),
                PendingIntent.FLAG_CANCEL_CURRENT);

        // Cancel all notifications in case the service may have been stopped by the system
        mNotificationManager.cancelAll();

        mDummyAlbumArt = BitmapFactory.decodeResource(service.getResources(),
                R.drawable.dummy_album_art);
        mAccentColor = ResourceHelper.getThemeColor(mService, R.attr.colorAccent, Color.DKGRAY);
    }

    public void startNotification() {
        if (!mStarted) {
            mMetadata = mController.getMetadata();
            mPlaybackState = mController.getPlaybackState();

            // Mise à jour de la notification après avoir mis mStarted à true
            Notification notif = createNotification();
            if (notif != null) {
                mController.registerCallback(mCallback);
                IntentFilter filter = new IntentFilter();
                filter.addAction(ACTION_NEXT);
                filter.addAction(ACTION_PAUSE);
                filter.addAction(ACTION_PLAY);
                filter.addAction(ACTION_PREV);
                mService.registerReceiver(this, filter);

                mService.startForeground(NOTIFICATION_ID, notif);
                mStarted = true;
            }
        }
    }

    private void stopNotification() {
        if (mStarted) {
            mStarted = false;
            mController.unregisterCallback(mCallback);
            try {
                mNotificationManager.cancel(NOTIFICATION_ID);
                mService.unregisterReceiver(this);
            } catch (IllegalArgumentException ex) {
                // On ignore le cas où le Receiver n'est pas enregistré.
            }
            mService.stopForeground(true);
        }
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        final String action = intent.getAction();
        switch (action) {
            case ACTION_PAUSE:
                mTransportControls.pause();
                break;
            case ACTION_PLAY:
                mTransportControls.play();
                break;
            case ACTION_NEXT:
                mTransportControls.skipToNext();
                break;
            case ACTION_PREV:
                mTransportControls.skipToPrevious();
                break;
            default:
                Log.w(TAG, "Unknown intent ignored. Action=" + action);
        }
    }

    private void updateSessionToken() {
        MediaSessionCompat.Token freshToken = mService.getSessionToken();
        if (freshToken != null && (mSessionToken == null || !mSessionToken.equals(freshToken))) {
            if (mController != null) {
                mController.unregisterCallback(mCallback);
            }
            try {
                mSessionToken = freshToken;
                mController = new MediaControllerCompat(mService, freshToken);
                mTransportControls = mController.getTransportControls();
                if (mStarted) {
                    mController.registerCallback(mCallback);
                }
            } catch (RemoteException | NullPointerException e) {
                Log.e(TAG, "updateSessionToken: error when creating MediaController.", e);
            }
        }
    }

    private PendingIntent createContentIntent() {
        Intent openUI = new Intent(mService, HomeActivity.class);
        openUI.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
        return PendingIntent.getActivity(mService, REQUEST_CODE, openUI,
                PendingIntent.FLAG_CANCEL_CURRENT);
    }

    private Notification createNotification() {
        if (mMetadata == null || mPlaybackState == null) {
            return null;
        }

        NotificationCompat.Builder notifBuilder = new NotificationCompat.Builder(mService);
        int playPauseButtonPosition = 0;

        // Si l'action "Previous" est activée
        if ((mPlaybackState.getActions()
                & PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS) != 0) {
            notifBuilder.addAction(R.drawable.ic_skip_previous_24dp,
                    mService.getString(R.string.action_previous), mPreviousIntent);
            // S'il y a un bouton "previous", alors "Play/Pause" est le second.
            // On garde ça en mémoire, parce que MediaStyle nécessite de préciser
            // l'index des boutons visibles dans le mode compact de la notification.
            playPauseButtonPosition = 1;
        }

        addPlayPauseButton(notifBuilder);

        // Si le bouton next est activé
        if ((mPlaybackState.getActions()
                & PlaybackStateCompat.ACTION_SKIP_TO_NEXT) != 0) {
            notifBuilder.addAction(R.drawable.ic_skip_next_24dp,
                    mService.getString(R.string.action_next), mNextIntent);
        }

        int smallIcon = mPlaybackState.getState() == PlaybackStateCompat.STATE_PLAYING
                ? R.drawable.notif_play_arrow
                : R.drawable.notif_pause;

        MediaDescriptionCompat description = mMetadata.getDescription();

        notifBuilder.setStyle(new NotificationCompat.MediaStyle()
                // N'affiche que le play/pause en mode compact
                .setShowActionsInCompactView(playPauseButtonPosition)
                .setMediaSession(mSessionToken))
                .setColor(mAccentColor)
                .setSmallIcon(smallIcon)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .setUsesChronometer(true)
                .setContentIntent(createContentIntent())
                .setContentTitle(description.getTitle())
                .setContentText(description.getSubtitle())
                .setLargeIcon(mDummyAlbumArt);

        setNotificationPlaybackState(notifBuilder);
        fetchAlbumArtAsync(description.getIconUri(), notifBuilder);

        return notifBuilder.build();
    }

    private void addPlayPauseButton(NotificationCompat.Builder builder) {
        String label;
        @DrawableRes int icon;
        PendingIntent intent;
        if (mPlaybackState.getState() == PlaybackStateCompat.STATE_PLAYING) {
            label = "Pause";
            icon = R.drawable.ic_pause;
            intent = mPauseIntent;
        } else {
            label = "Play";
            icon = R.drawable.ic_play_arrow;
            intent = mPlayIntent;
        }
        builder.addAction(new NotificationCompat.Action(icon, label, intent));
    }

    private void setNotificationPlaybackState(NotificationCompat.Builder builder) {
        Log.d(TAG, "setNotificationPlaybackState. mPlaybackState=" + mPlaybackState);
        if (mPlaybackState == null || !mStarted) {
            Log.d(TAG, "setNotificationPlaybackState: cancelling notification!");
            mService.stopForeground(true);
            return;
        }

        if (mPlaybackState.getState() == PlaybackStateCompat.STATE_PLAYING
                && mPlaybackState.getPosition() >= 0) {
            Log.d(TAG, "setNotificationPlaybackState. updating playback position");
            builder.setWhen(System.currentTimeMillis() - mPlaybackState.getPosition())
                    .setShowWhen(true)
                    .setUsesChronometer(true);
        } else {
            Log.d(TAG, "setNotificationPlaybackState. hiding playback position");
            builder.setWhen(0)
                    .setShowWhen(false)
                    .setUsesChronometer(false);
        }

        // Permet de swiper la notification lorsque le player est en pause
        builder.setOngoing(mPlaybackState.getState() == PlaybackStateCompat.STATE_PLAYING);
    }

    private void fetchAlbumArtAsync(final Uri artUri, final NotificationCompat.Builder builder) {
        Drawable placeholder = null;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            // FIXME Pre-lollipop don't like Vector whatever I do
            placeholder = AppCompatResources.getDrawable(mService.getApplicationContext(), R.drawable.ic_audiotrack_24dp);
        }
        Glide.with(mService.getApplicationContext())
                .load(artUri).asBitmap()
                .error(placeholder)
                .diskCacheStrategy(DiskCacheStrategy.NONE)
                .into(new SimpleTarget<Bitmap>() {
                    @Override
                    public void onResourceReady(Bitmap art, GlideAnimation<? super Bitmap> anim) {
                        // Si la pochette d'album est toujours celle du titre en cours de lecture
                        if (artUri.equals(mMetadata.getDescription().getIconUri())) {
                            Log.d(TAG, "fetchAlbumArtAsync: set bitmap to " + artUri);
                            builder.setLargeIcon(art);
                            mNotificationManager.notify(NOTIFICATION_ID, builder.build());
                        }
                    }
                });
    }
}

File: app/src/main/java/fr/nihilus/mymusic/MediaNotificationManager.kt
package fr.nihilus.mymusic

import android.app.Notification
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.BitmapFactory
import android.graphics.Color
import android.os.RemoteException
import android.support.annotation.DrawableRes
import android.support.v4.app.NotificationCompat
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.MediaControllerCompat
import android.support.v4.media.session.MediaSessionCompat
import android.support.v4.media.session.PlaybackStateCompat
import android.util.Log
import fr.nihilus.mymusic.di.MusicServiceScope
import fr.nihilus.mymusic.utils.ResourceHelper
import javax.inject.Inject

@MusicServiceScope
internal class MediaNotificationManager
@Inject constructor(service: MusicService) : BroadcastReceiver() {

    private val mService = service

    private val mNotificationManager: NotificationManager
    private val mPauseIntent: PendingIntent
    private val mPlayIntent: PendingIntent
    private val mPreviousIntent: PendingIntent
    private val mNextIntent: PendingIntent
    private val mNotificationColor = ResourceHelper.getThemeColor(service, R.attr.colorPrimary, Color.DKGRAY)

    private var mController: MediaControllerCompat? = null
    private var mTransportControls: MediaControllerCompat.TransportControls? = null

    private var mSessionToken: MediaSessionCompat.Token? = null
    private var mPlaybackState: PlaybackStateCompat? = null

    private var mMetadata: MediaMetadataCompat? = null
    private var mStarted = false

    init {
        val pkg = service.packageName
        updateSessionToken()

        mPauseIntent = PendingIntent.getBroadcast(service, REQUEST_CODE,
                Intent(ACTION_PAUSE).setPackage(pkg), PendingIntent.FLAG_CANCEL_CURRENT)
        mPlayIntent = PendingIntent.getBroadcast(service, REQUEST_CODE,
                Intent(ACTION_PLAY).setPackage(pkg), PendingIntent.FLAG_CANCEL_CURRENT)
        mPreviousIntent = PendingIntent.getBroadcast(service, REQUEST_CODE,
                Intent(ACTION_PREVIOUS).setPackage(pkg), PendingIntent.FLAG_CANCEL_CURRENT)
        mNextIntent = PendingIntent.getBroadcast(service, REQUEST_CODE,
                Intent(ACTION_NEXT).setPackage(pkg), PendingIntent.FLAG_CANCEL_CURRENT)

        // Cancel all notifications to handle the case where the Service was killed
        // and restarted by the system.
        mNotificationManager = service.getSystemService(Context.NOTIFICATION_SERVICE)
                as NotificationManager
        mNotificationManager.cancelAll()
    }

    private fun updateSessionToken() {
        val freshToken = mService.sessionToken
        if (mSessionToken == null && freshToken != null ||
                mSessionToken != null && mSessionToken != freshToken) {
            mController?.unregisterCallback(mControllerCallback)
            mSessionToken = freshToken
            if (mSessionToken != null) {
                mController = MediaControllerCompat(mService, mSessionToken!!)
                mTransportControls = mController!!.transportControls
                if (mStarted) {
                    mController!!.registerCallback(mControllerCallback)
                }
            }
        }
    }

    fun startNotification() {
        if (!mStarted) {
            mMetadata = mController!!.metadata
            mPlaybackState = mController!!.playbackState

            val notification = createNotification()
            if (notification != null) {
                mController!!.registerCallback(mControllerCallback)
                val filter = IntentFilter()
                filter.addAction(ACTION_PAUSE)
                filter.addAction(ACTION_PLAY)
                filter.addAction(ACTION_PREVIOUS)
                filter.addAction(ACTION_NEXT)
                mService.registerReceiver(this, filter)

                mService.startForeground(NOTIFICATION_ID, notification)
                mStarted = true
            }
        }
    }

    override fun onReceive(ctx: Context, intent: Intent) {
        val action = intent.action
        Log.v(TAG, "Received intent with action $action")
        when (action) {
            ACTION_PAUSE -> mTransportControls!!.pause()
            ACTION_PLAY -> mTransportControls!!.play()
            ACTION_PREVIOUS -> mTransportControls!!.skipToPrevious()
            ACTION_NEXT -> mTransportControls!!.skipToNext()
            else -> Log.w(TAG, "Unknown intent ignored. Action=$action")
        }
    }

    private val mControllerCallback = object : MediaControllerCompat.Callback() {

        override fun onPlaybackStateChanged(state: PlaybackStateCompat) {
            mPlaybackState = state
            Log.v(TAG, "Received new playback state $state")
            if (state.state == PlaybackStateCompat.STATE_STOPPED ||
                    state.state == PlaybackStateCompat.STATE_NONE) {
                stopNotification()
            } else {
                val notification = createNotification()
                if (notification != null) {
                    mNotificationManager.notify(NOTIFICATION_ID, notification)
                }
            }
        }

        override fun onMetadataChanged(metadata: MediaMetadataCompat?) {
            mMetadata = metadata
            Log.v(TAG, "Received new metadata $metadata")
            val notification = createNotification()
            if (notification != null) {
                mNotificationManager.notify(NOTIFICATION_ID, notification)
            }
        }

        override fun onSessionDestroyed() {
            Log.v(TAG, "Session was destroyed, resetting to the new session token")
            try {
                updateSessionToken()
            } catch (e: RemoteException) {
                Log.e(TAG, "Could not connect to MediaController", e)
            }
        }
    }

    private fun createContentIntent(): PendingIntent {
        val openUi = Intent(mService, HomeActivity::class.java)
        openUi.flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
        return PendingIntent.getActivity(mService, REQUEST_CODE, openUi,
                PendingIntent.FLAG_CANCEL_CURRENT)
    }

    private fun createNotification(): Notification? {
        Log.v(TAG, "updateNotificationMetadata: metadata=$mMetadata")
        if (mMetadata == null || mPlaybackState == null) return null

        val notificationBuilder = NotificationCompat.Builder(mService)
        notificationBuilder.addAction(R.drawable.ic_skip_previous_24dp,
                mService.getString(R.string.action_previous), mPreviousIntent)

        addPlayPauseButton(notificationBuilder)

        notificationBuilder.addAction(R.drawable.ic_skip_next_24dp,
                mService.getString(R.string.action_next), mNextIntent)

        // TODO Maybe replace by the description calculated by MusicRepository
        val description = mMetadata!!.description

        // TODO handle album art loading (placeholder used instead)
        val albumArt = BitmapFactory.decodeResource(mService.resources,
                R.drawable.dummy_album_art)

        val smallIcon = if (mPlaybackState!!.state == PlaybackStateCompat.STATE_PLAYING)
            R.drawable.ic_pause else R.drawable.ic_play_arrow

        notificationBuilder.setStyle(android.support.v4.media.app.NotificationCompat.MediaStyle()
                .setShowActionsInCompactView(1)
                .setMediaSession(mSessionToken))
                .setColor(mNotificationColor)
                .setSmallIcon(smallIcon)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .setUsesChronometer(true)
                .setContentIntent(createContentIntent())
                .setContentTitle(description.title)
                .setContentText(description.subtitle)
                .setLargeIcon(albumArt)

        setNotificationPlaybackState(notificationBuilder)
        return notificationBuilder.build()
    }

    private fun addPlayPauseButton(builder: NotificationCompat.Builder) {
        Log.v(TAG, "updatePlayPauseAction")
        val label: String
        val intent: PendingIntent
        @DrawableRes val icon: Int

        if (mPlaybackState!!.state == PlaybackStateCompat.STATE_PLAYING) {
            label = "Pause"
            icon = R.drawable.ic_pause
            intent = mPauseIntent
        } else {
            label = "Play"
            icon = R.drawable.ic_play_arrow
            intent = mPlayIntent
        }

        builder.addAction(icon, label, intent)
    }

    private fun setNotificationPlaybackState(builder: NotificationCompat.Builder) {
        Log.v(TAG, "updateNotificationPlaybackState: mPlaybackState=$mPlaybackState")
        if (mPlaybackState == null || !mStarted) {
            Log.v(TAG, "updatedNotificationPlaybackState: canceling notification!")
            mService.stopForeground(true)
            return
        }

        if (mPlaybackState!!.state == PlaybackStateCompat.STATE_PLAYING
                && mPlaybackState!!.position >= 0) {
            builder.setWhen(System.currentTimeMillis() - mPlaybackState!!.position)
                    .setShowWhen(true)
                    .setUsesChronometer(false)
        } else {
            Log.v(TAG, "updateNotificationPlaybackState: hiding playback position")
            builder.setWhen(0)
                    .setShowWhen(false)
                    .setUsesChronometer(false)
        }

        // Make sure that the notification can be dismissed by the user when we are not playing
        builder.setOngoing(mPlaybackState!!.state == PlaybackStateCompat.STATE_PLAYING)
    }

    private fun stopNotification() {
        if (mStarted) {
            mStarted = false
            mController!!.unregisterCallback(mControllerCallback)
            try {
                mNotificationManager.cancel(NOTIFICATION_ID)
                mService.unregisterReceiver(this)
            } catch (e: IllegalArgumentException) {
                // Ignore if the receiver is not registered.
            } finally {
                mService.stopForeground(true)
            }
        }
    }

    companion object {
        private const val TAG = "MediaNotifMgr"
        const val REQUEST_CODE = 100
        const val NOTIFICATION_ID = 42
        const val ACTION_PLAY = "fr.nihilus.music.action.PLAY"
        const val ACTION_PAUSE = "fr.nihilus.music.action.PAUSE"
        const val ACTION_PREVIOUS = "fr.nihilus.music.action.PREVIOUS"
        const val ACTION_NEXT = "fr.nihilus.music.action.NEXT"
    }
}